src/mpid/ch4/src/ch4r_callbacks.c

```diff
@@ -472,7 +474,7 @@ int MPIDIG_send_data_target_msg_cb(int handler_id, void *am_hdr, void *data, MPI
                     (MPL_DBG_FDEST, "seq_no: me=%" PRIu64 " exp=%" PRIu64,
                      MPIDIG_REQUEST(rreq, req->seq_no),
                      MPL_atomic_load_uint64(&MPIDI_global.exp_seq_no)));
-    MPIDIG_recv_type_init(in_data_sz, rreq);
+    MPIDIG_recv_type_init(MPIDIG_REQUEST(rreq, req->rreq.data_sz), rreq);
```

Comment:
Is this is the same `rreq`? Why not simply always do a `MPIDIG_recv_type_init` earlier in `MPIDIG_send_target_msg_cb` and skip line 436 and here?

Reply:
It kind of hard because moving it up into `MPIDIG_send_target_msg_cb` is making the code init the recv before the recv datatype/count available (for unexpected). In that case, we would need to reinit when the matching recv is posted.